### PR TITLE
[clang] Fix incorrect partial ordering context setting

### PR DESCRIPTION
**Pull request by mizvekov**
Fixes regression introduced in #94981, reported on the pull-request.

Since this fixes a commit which was never released, there are no release notes.

### CODE REVIEW

File: clang/lib/Sema/TreeTransform.h [hunk 1]
```diff
@@ -4661,6 +4718,42 @@ TreeTransform<Derived>::TransformTemplateName(CXXScopeSpec &SS,
...
+        SS, Underlying, NameLoc, ObjectType, FirstQualifierInScope,
+        AllowInjectedClassName);
+    if (TransUnderlying.isNull())
+      return TemplateName();
+
+    DefaultArguments DefArgs = DTN->getDefaultArguments();
+
+    TemplateArgumentListInfo TransArgsInfo;
+    using Iterator =
+        TemplateArgumentLocInventIterator<Derived, TemplateArgument *>;
```
Comment (cor3ntin) [inline clang/lib/Sema/TreeTransform.h:4733]:
```suggestion
    using Iterator =
        TemplateArgumentLocInventIterator<Derived, const TemplateArgument *>;
```

I think you might be able to get rid of the `const_cast`

**Comment by alexfh** [comment]
Unfortunately, the original code, from which I reduced the test case for this, still breaks (without crashing Clang though). I'm trying to come up with another test case now.

**Comment by alexfh** [comment]
I could reduce the code to something that compiles with Clang before fa6580470547411667b866362941db0b02e25578, but not after it, this PR doesn't fix the issue. The example is: https://gcc.godbolt.org/z/odWYhxGxK

Note that the code doesn't compile with Clang 18.1, but should compile with Clang 19 (which I couldn't find on Compiler Explorer) and for that matter with all Clang versions starting from b86e0992bfa6c58be077d82d824016f590ac5d90 and up to fa6580470547411667b866362941db0b02e25578.

**Comment by mizvekov** [comment]
@alexfh thanks, the latest push should fix it.

**Review by alexfh — APPROVED** [review]
Thanks! This fixes the issues we've found so far, and it seems to be a quite clear correctness fix. LGTM

**Review by cor3ntin — APPROVED** [review]

**Pull request merged**